common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java

```diff
@@ -1267,6 +1268,128 @@ public static UTF8String[] icuSplitSQL(final UTF8String string, final UTF8String
+      // that we will process.
+      newWord = (codepoint == ASCII_SPACE_CODEPOINT);
+      if(!UCharacter.hasBinaryProperty(codepoint, UProperty.CASE_IGNORABLE)) {
+        precededByCasedLetter = UCharacter.hasBinaryProperty(codepoint, UProperty.CASED);
+      }
+      offset += UTF8String.numBytesForFirstByte(target.getByte(offset));
+    }
+    return UTF8String.fromString(sb.toString());
+  }
+
+  private static void appendTitleCasedCodepoint(
+      StringBuilder sb,
+      int codepoint,
+      boolean isAfterAsciiSpace,
+      boolean precededByCasedLetter,
+      UTF8String target,
+      int offset) {
+    if(isAfterAsciiSpace) {
+      // Titlecasing a character if it is in the beginning of a new word.
+      sb.append(codepointToTitleString(codepoint));
+      return;
+    }
+    if(codepoint == CAPITAL_SIGMA) {
+      // Handling capital Greek letter sigma ('Σ').
+      appendLowerCasedGreekCapitalSigma(sb, precededByCasedLetter, target, offset);
+      return;
+    }
+    // If it's not the beginning of a word, or a capital Greek letter sigma ('Σ'), we lowercase the
+    // character.
+    sb.append(
+      toLowerCase(UTF8String.fromString(new String(Character.toChars(codepoint)))).toString());
+  }
+
+  private static void appendLowerCasedGreekCapitalSigma(
+      StringBuilder sb,
+      boolean precededByCasedLetter,
+      UTF8String target,
+      int offset) {
+    int codepoint;
+    if (!followedByCasedLetter(target,offset) && precededByCasedLetter) {
+      codepoint = SMALL_FINAL_SIGMA;
+    }
+    else {
+      codepoint = SMALL_NON_FINAL_SIGMA;
+    }
+    sb.appendCodePoint(codepoint);
+  }
+
+  /**
+   * Checks if the character beginning at 'offset'(in 'targets' byte array) is followed by a cased
+   * letter.
+   *
+   * @param target
+   * @param offset
+   * @return
+   */
+  private static boolean followedByCasedLetter(UTF8String target, int offset) {
+    // Moving the offset one character forward, so we could start the linear search from there.
+    offset += UTF8String.numBytesForFirstByte(target.getByte(offset));
+    int len = target.numBytes();
+
+    while(offset < len) {
+      int codepoint = target.codePointFrom(offset);
+
+      if(UCharacter.hasBinaryProperty(codepoint, UProperty.CASE_IGNORABLE)) {
+        offset += UTF8String.numBytesForFirstByte(target.getByte(offset));
+        continue;
+      }
+      return UCharacter.hasBinaryProperty(codepoint, UProperty.CASED);
+    }
+    return false;
+  }
+
+  /**
+   * Titlecases a single character using the ICU root locale rules.
+   *
+   * @param codepoint is a character that is needed to be title cased
+   * @return a java String(whose length can be more than 1 character) which corresponds to the title
+   * case of the codepoint
+   */
+  private static String codepointToTitleString(int codepoint) {
+    return UCharacter.toTitleCase(new String(Character.toChars(codepoint)),null);
```

Comment:
let's just add a comment here explaining that this is a generally expensive operation for a single code point

it is my understanding that, in the worst case, we can have O(n/2) of these operations (where n is the length of the input string that we're titlecasing)